src/Microsoft.Extensions.Primitives/CompositeChangeToken.cs

```diff
@@ -0,0 +1,165 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.Extensions.Primitives
+{
+    /// <summary>
+    /// An <see cref="IChangeToken"/> which represents one or more <see cref="IChangeToken"/> instances.
+    /// </summary>
+    public class CompositeChangeToken : IChangeToken
+    {
+        private readonly object _callbackLock = new object();
+        private List<CallbackState> _callbacks = new List<CallbackState>();
+        private bool _registeredCallbackProxy;
+        private bool _canBeChanged = true;
+
+        /// <summary>
+        /// Creates a new instance of <see cref="CompositeChangeToken"/>.
+        /// </summary>
+        /// <param name="changeTokens">The list of <see cref="IChangeToken"/> to compose.</param>
+        public CompositeChangeToken(IReadOnlyList<IChangeToken> changeTokens)
+        {
+            if (changeTokens == null)
+            {
+                throw new ArgumentNullException(nameof(changeTokens));
+            }
+
+            ChangeTokens = changeTokens;
+        }
+
+        /// <summary>
+        /// Returns the list of <see cref="IChangeToken"/> which compose the current <see cref="CompositeChangeToken"/>.
+        /// </summary>
+        public IReadOnlyList<IChangeToken> ChangeTokens { get; }
+
+        /// <inheritdoc />
+        public virtual IDisposable RegisterChangeCallback(Action<object> callback, object state)
+        {
+            if (!_canBeChanged)
+            {
+                return NullDisposable.Singleton;
+            }
+
+            var changeTokenState = new CallbackState(callback, state, new CallbackDisposable());
```

Comment:
Would that cancel the callbacks for individual tokens? Instead of disposing CallbackState objects, you're suggesting that we add them to our list of callbacks and cancel them?